compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -146,6 +147,50 @@ impl<'tcx> Cx<'tcx> {
                 ExprKind::RawBorrow { mutability, arg: self.thir.exprs.push(expr) }
             }
             Adjust::DynStar => ExprKind::Cast { source: self.thir.exprs.push(expr) },
+            Adjust::ReborrowPin(AutoBorrow::Ref(region, m)) => {
```

Comment:
I think lowering to `as_mut` might be problematic. Specifically, because of the `Deref` bound and the "generality" of `as_mut`, this code breaks:

```
#![feature(pin_ergonomics)]

use std::pin::Pin;

fn shorter<'b, T: 'b>(x: Pin<&'b mut T>) {}

fn test<'a: 'b, 'b, T: 'a>(mut x: Pin<&'a mut T>) {
    shorter::<'b>(x);
}
```

Because `Pin::as_mut` takes a ref to the *pin*. I think you want to do some sort of desugaring that is more analogous to:

```
fn shorten<'a: 'b, 'b, T>(_: Pin<&'a mut T>) -> Pin<&'b mut T>
```

Reply:
Okay, I just pushed some changes that desugar the reborrow as `Pin { __pointer: &mut *x.__pointer }` which solves this issue.